### PR TITLE
Introduce torch_remat activation checkpointing 

### PR DESCRIPTION
**Pull request by anijain2305**
Human note heavily polished by AI

This PR focusses on the UX side of things with minimal model code surgery. The goal of this PR is to agree on the design. After this PR, the  changes in next PRs are lot more mechanical or focus on improving debug experience.

Motivation
----------

TorchTitan currently provides full and selective activation checkpointing. SelectiveAC makes save decisions at the operator level. torch_remat provides a model-aware alternative where model code defines semantic regions and the training configuration chooses which region outputs to retain.

This requires small, explicit annotations in model code. That tradeoff makes the available policy surface visible next to the operations it controls and allows policies to refer to stable model concepts rather than individual ATen operators.

User experience
---------------

RematAC applies remat.checkpoint to each transformer layer under model.layers. Model components advertise optional save boundaries with AVAILABLE_REMAT_SAVE_REGIONS and call remat.region at those boundaries.

Users select regions through RematAC.Config:

    RematAC.Config(
        save_regions=[
            "attention.qkv",
            "attention.wo",
        ]
    )

Region names are qualified relative to one transformer layer. The same policy therefore applies to every layer. Shell-style patterns are supported, so "attention.*" selects every region exposed by the attention module. Overlapping patterns are a set union because every pattern expresses the same SAVE action.

An empty save_regions list is valid. It recomputes all user-configurable regions. A pattern that does not match the current model part raises an actionable error by default. Pipeline stages that intentionally do not contain a matching region can set allow_unmatched_save_regions=True; the unmatched patterns still produce a warning.

Attention example
-----------------

GQAttention exposes three regions:

    AVAILABLE_REMAT_SAVE_REGIONS = (
        "qkv",
        "inner_attention",
        "wo",
    )

For the configuration above, execution is equivalent to:

    remat.checkpoint(transformer_layer)
    |
    +-- attention
        |
        +-- qkv projections ............. SAVE
        +-- q_norm / k_norm ............. RECOMPUTE
        +-- RoPE ........................ RECOMPUTE
        +-- inner_attention ............. RECOMPUTE
        +-- output reshape .............. RECOMPUTE
        +-- wo .......................... SAVE

SAVE means that the region runs during the original forward and its output is retained for backward. RECOMPUTE means that the operation runs again during checkpoint replay.

Implementation
--------------

Model call sites use torch_remat directly:

  * remat.region receives the configured qualified name and recompute decision.
  * remat.recompute_needs_tensor marks saved outputs needed by recomputed bare operations outside another region.
  * Module.remat_region_name and Module.remat_should_recompute provide safe defaults, so region annotations do not change execution without an enclosing remat checkpoint.

configure_remat_save_regions walks one transformer layer, discovers declared regions, matches the configured patterns, and installs names and decisions on the participating module instances. A subclass inherits a region declaration only while it also inherits the associated forward implementation. A subclass that replaces forward must redeclare and implement its own regions.

The MoE routing decision is handled separately from user policy. Node-limited selection and final topk run inside an internal recompute=False region because replaying tied topk values can select different experts. This correctness boundary is intentionally not exposed through AVAILABLE_REMAT_SAVE_REGIONS.

torch_remat is promoted from an optional dependency to a regular TorchTitan dependency because common model code calls its APIs directly.

Supported in this change
------------------------

  * Transformer block containers named model.layers.
  * Common GQAttention implementations that use the shared forward method.
  * qkv, inner_attention, and wo as independently selectable save regions.
  * Exact region names and shell-style wildcard patterns.
  * Empty policies that recompute every user-configurable region.
  * Stage-local pipeline application through the existing activation- checkpointing flow, with an explicit escape hatch for stage-local unmatched patterns.

Not supported yet
-----------------

  * Different policies for individual transformer layers.
  * Block containers other than model.layers.
  * User-selectable FFN, expert, dispatch, vision, or model-specific attention regions.
  * Attention subclasses that replace the common GQAttention forward without declaring their own region boundaries.
  * Global unmatched-pattern validation across all pipeline stages.
  * Activation-checkpoint debug mode or implicit RNG preservation inside saved regions. Random state requires explicit torch_remat state hooks.
  * Moving TP/SP source-to-destination redistributions from generic module wrappers into the corresponding model regions.

Validation
----------

Focused unit tests verify wildcard selection, empty policies, unmatched-pattern errors and the pipeline escape hatch, forward/backward equality, per-region execution counts, inherited-forward handling, state-dict stability, and the always-saved routing decision.

The model integration suite adds a Llama 3 FSDP2+TP2+CP2 RematAC configuration and compares loss and gradient norm against the existing Llama numerical golden.

### CODE REVIEW

File: .ci/docker/common/install_conda.sh [hunk 1]
```diff
@@ -47,6 +47,8 @@ install_pip_dependencies() {
   # --no-deps: the image ships no torch on purpose, every job installs a nightly.
   pip_install --no-deps \
     "git+https://github.com/meta-pytorch/torch_checkpointing.git@main"
+  pip_install --no-deps \
+    "git+https://github.com/meta-pytorch/remat.git@main"
```
Comment (tianyu-l) [inline .ci/docker/common/install_conda.sh]: no need if you already put it in requirements.txt (line 43)

File: tests/unit_tests/gpu/test_remat.py
Comment (tianyu-l) [inline tests/unit_tests/gpu/test_remat.py:1]: cpu/ or gpu/

File: torchtitan/distributed/activation_checkpoint.py [hunk 2]
```diff
@@ -292,6 +294,125 @@ def wrapped_policy(ctx, func, *args, **kwargs) -> CheckpointPolicy:
         )
 
 
+class RematAC(ActivationCheckpointing):
```
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py]: `RematAC` reads as rematerialize activation checkpointing, which doesn't sound like a term.

Given that we are in the process of migrating, I don't think it matters that much. But at least add a note that we're going to remove all other AC variants and only keep this, at which point it deserves a serious name.

File: pyproject.toml [hunk 3]
```diff
@@ -77,5 +75,5 @@ markers = [
 [tool.pyrefly]
 python-version = "3.11"
 project-excludes = ["torchtitan/experiments", "**/tests/**"]
-replace-imports-with-any = ["torchao.*", "torchft", "torchvision.*", "deep_ep.*", "jinja2.*", "helion", "helion.*", "batch_invariant_ops", "torchcomms"]  # optional dependencies
+replace-imports-with-any = ["torchao.*", "torchft", "torchvision.*", "deep_ep.*", "jinja2.*", "helion", "helion.*", "batch_invariant_ops", "torchcomms", "torch_remat"]  # dependencies without Pyrefly stubs
```
Comment (tianyu-l) [inline pyproject.toml]: since it's not optional, shouldn't be put in this list?

File: torchtitan/distributed/activation_checkpoint.py
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py:1]: I think we could benefit from an overall REMAT.md as tutorial for torchtitan users / agents.

File: torchtitan/models/common/attention.py [hunk 4]
```diff
@@ -942,13 +954,25 @@ def forward(
...
+        )(
             xq_THK,
             xk_THK,
             xv_THV,
             attention_masks=attention_masks,
             scale=self.scaling,
             enable_gqa=self.enable_gqa,
-        ).contiguous()
+        )
+        remat.recompute_needs_tensor(out_THV)
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:963]: we need a note somewhere to explicitly mention this
> During forwards, we try to infer whether or not an output of a SAVE region will need to be saved by tracking if it ends up being used by a RECOMPUTE region. If you have an explicit `remat.region(..., recompute=True)` that consumes this tensor, we can infer it automatically; otherwise, you should call `remat.recompute_needs_tensor` write before its usage

Comment (tianyu-l) [inline torchtitan/models/common/attention.py:963]: If above region is recomputed, but user didn't specify this, would it error out? What error would user see?

File: torchtitan/distributed/activation_checkpoint.py [hunk 5]
```diff
@@ -292,6 +294,125 @@ def wrapped_policy(ctx, func, *args, **kwargs) -> CheckpointPolicy:
...
+        Unmatched patterns still emit a warning.
+        """
+
+        preserve_rng_state: bool = False
+        """
+        Must remain false. torch_remat requires explicit RecomputeStateHooks for
+        random state that can advance inside retained regions.
+        """
+
+    def _validate_config(self) -> "RematAC.Config":
```
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py]: for now let's put in post_init of RematAC.Config

File: torchtitan/models/common/attention.py [hunk 6]
```diff
@@ -869,6 +870,12 @@ class GQAttention(BaseAttention):
     :class:`FusedQKVLinear` for a single fused projection.
     """
 
+    AVAILABLE_REMAT_SAVE_REGIONS: tuple[str, ...] = (
+        "qkv",
+        "inner_attention",
+        "wo",
+    )
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: remove for now

File: torchtitan/distributed/activation_checkpoint.py [hunk 7]
```diff
@@ -292,6 +294,125 @@ def wrapped_policy(ctx, func, *args, **kwargs) -> CheckpointPolicy:
...
+            )
+        )
+        unmatched_patterns = [
+            pattern
+            for pattern in config.save_regions
+            if not any(fnmatch(region, pattern) for region in available_regions)
+        ]
+        # This model can be one pipeline partition, so a pattern absent here may
+        # still match a region assigned to another stage.
+        if unmatched_patterns:
```
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py]: let's remove this check for now -- in first version let's focus on a bare minimal version of integration

File: torchtitan/distributed/activation_checkpoint.py [hunk 8]
```diff
@@ -292,6 +294,125 @@ def wrapped_policy(ctx, func, *args, **kwargs) -> CheckpointPolicy:
...
+                    "allow_unmatched_save_regions=True."
+                )
+            logger.warning(
+                "RematAC save_regions patterns did not match this model part: %s",
+                unmatched_patterns,
+            )
+
+        selected_regions = []
+        for layer_id, transformer_block in transformer_blocks:
+            selected_for_block, _ = configure_remat_save_regions(
```
Comment (tianyu-l) [inline torchtitan/distributed/activation_checkpoint.py]: Instead of introducing this function, could you directly do it inside `module.configure_remat_regions`?

File: torchtitan/models/common/attention.py (hunk 6 above)
Comment (anijain2305) [reply torchtitan/models/common/attention.py]: Do you mean to remove `AVAILABLE_REMAT_SAVE_REGIONS` or entire changes in `torchtitan/models/common/attention.py`? I deliberately kept this file change so that we can discuss the level of surgery and interface that future PRs will adopt.

File: torchtitan/models/common/attention.py (hunk 4 above)
Comment (anijain2305) [reply torchtitan/models/common/attention.py:963]: The backward will error out during the forward recomputation. 	

```

  import torch
  import torch_remat as remat


  def body(x):
      y = remat.region(lambda t: t * 2, "saved_mul", recompute=False)(x)
      return torch.relu(y)  # Bare consumer; marker intentionally omitted.


  x = torch.tensor([1.0, -1.0], requires_grad=True)
  y = remat.checkpoint(region_name="block")(body)(x)
  y.sum().backward()
```
fails with a Runtime Error - full trace here - https://www.internalfb.com/phabricator/paste/view/P2493483097

In TorchTitan, currently my plan is to add it everywhere after remat.region because the region outputs are way more commonly consumed right away. Only when its the last region of the transfomer remat.checkpoint boundary, we should not store it. I plan to look at this later.

File: pyproject.toml (hunk 3 above)
Comment (anijain2305) [reply pyproject.toml]: This is for pyrefly stubs handling.

File: torchtitan/distributed/activation_checkpoint.py
Comment (anijain2305) [reply torchtitan/distributed/activation_checkpoint.py:1]: Added - docs/remat.md

File: torchtitan/models/common/attention.py (hunk 4 above)
Comment (anijain2305) [reply torchtitan/models/common/attention.py:963]: Added in remat.md

File: pyproject.toml [hunk 9]
```diff
@@ -29,6 +29,8 @@ dependencies = [
     "pillow",
     "spmd_types==0.2.5",
     "attn-gym[linear]==0.0.8",
+    # Keep this commit pin in sync with requirements.txt.
+    "torch_remat @ git+https://github.com/meta-pytorch/remat.git@d302699b1c58f83fa2c7b03bc2593967e9530335",
```
Comment (anijain2305) [inline pyproject.toml:33]: Unfortunately, we will have to work with pins  until the integration is stable and we can cut a torch_remat release.

File: torchtitan/models/common/attention.py (hunk 6 above)
Comment (anijain2305) [reply torchtitan/models/common/attention.py]: I removed the available tuple. It sat well with the minimal api in the rPR

File: torchtitan/models/common/attention.py [hunk 10]
```diff
@@ -931,7 +932,12 @@ def forward(
         attention_masks: AttentionMasksType | None,
         positions: torch.Tensor | None = None,
     ) -> torch.Tensor:
-        xq_THK, xk_THK, xv_THV = self.qkv_linear(x_TD)
+        xq_THK, xk_THK, xv_THV = remat.region(
+            self.qkv_linear,
+            self.remat_region_name("qkv"),
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: Do we need this because we need module name as prefix to avoid collision? E.g. if both MLP and Attention has `w1`, specifying `"w1"` here would cause collision, so we need `"attention.w1"`

**Review by tianyu-l — APPROVED** [review]
very cool
